### PR TITLE
added doctor for ward

### PR DESCRIPTION
**Pull request by sviridchik**

### CODE REVIEW

**Review by AHTOH2001 — CHANGES_REQUESTED** [review]
Merge `DoctorForWardViewSet` with `DoctorVisitViewSet` and distinguish users by `request.user.guardian` or `request.user.patient`

**Comment by sviridchik** [comment]
h

File: api/api/settings.py [hunk 1]
```diff
@@ -91,22 +91,22 @@
...
+#         'HOST': '127.0.0.1',
+#         'PORT': 5432,
 #     }
 # }
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3',
+        'NAME': BASE_DIR / 'db.sqlite3',
+    }
+}
```
Comment (AHTOH2001) [inline api/api/settings.py]: Revert postgres

File: api/managment/views.py [hunk 2]
```diff
@@ -97,14 +89,30 @@ class TranzactionViewSet(viewsets.ModelViewSet):
 
 class DoctorViewSet(viewsets.ModelViewSet):
     serializer_class = DoctorSerializer
-    permission_classes = [IsAuthenticated]
+    # permission_classes = [IsAuthenticated]
```
Comment (AHTOH2001) [inline api/managment/views.py]: Uncomment this

File: api/managment/views.py [hunk 3]
```diff
@@ -3,21 +3,13 @@
 from rest_framework.decorators import api_view, permission_classes
 from rest_framework.permissions import IsAuthenticated
 from rest_framework.response import Response
+from rest_framework import status
+from  medicine.views import get_type_of_user
+from .models import Patient, PatientSetting, Guardian, Tariff, Tokens, Tranzaction, Doctor, DoctorVisit
+from .serializers import PatientSerializer, PatientSettingSerializer, GuardianSerializer, TariffSerializer, \
+    TokensSerializer, TranzactionSerializer, DoctorVisitSerializer, DoctorSerializer, UserSerializer, ReadOnlyDoctorVisitSerializer
+from rest_framework.permissions import IsAuthenticated
```
Comment (AHTOH2001) [inline api/managment/views.py:22]: Wrong imports formatiing

File: api/medicine/views.py [hunk 4]
```diff
@@ -112,28 +112,48 @@ def get(self, request, pk, *args, **kwargs):
...
+        # self.request.user.patient
+        data.guardian
+    except Guardian.DoesNotExist as e:
+        try:
+            # self.request.user.patient
+            data.patient
+        except Patient.DoesNotExist as e:
+            return "not found"
+        return "patient"
+    return "guardian"
```
Comment (AHTOH2001) [inline api/medicine/views.py]: Create file api/managment/utils.py and put this method there

File: api/medicine/views.py [hunk 5]
```diff
@@ -112,28 +112,48 @@ def get(self, request, pk, *args, **kwargs):
             return Response({"error": "no need to take it"}, status=status.HTTP_400_BAD_REQUEST)
         return Response(serializer.data)
 
-
+def get_type_of_user(data):
```
Comment (AHTOH2001) [inline api/medicine/views.py]: Rename data to user

File: api/medicine/views.py [hunk 6]
```diff
@@ -112,28 +112,48 @@ def get(self, request, pk, *args, **kwargs):
...
+    try:
+        # self.request.user.patient
```
Comment (AHTOH2001) [inline api/medicine/views.py]: Do not write redundant comments

**Review by AHTOH2001 — CHANGES_REQUESTED** [review]

File: api/medicine/views.py (hunk 4 above)
Comment (AHTOH2001) [inline api/medicine/views.py]: Rewrite this method by analogy of whoiam

File: api/managment/views.py [hunk 7]
```diff
@@ -97,14 +89,30 @@ class TranzactionViewSet(viewsets.ModelViewSet):
...
 class DoctorViewSet(viewsets.ModelViewSet):
     serializer_class = DoctorSerializer
-    permission_classes = [IsAuthenticated]
+    # permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        type_user = get_type_of_user(self.request.user)
+        if type_user == "guardian":
+            try:
+                ward = int(self.request.GET.get('ward'))
```
Comment (AHTOH2001) [inline api/managment/views.py]:
```suggestion
                ward = int(self.request.query_params['ward'])
```

File: api/managment/views.py [hunk 8]
```diff
@@ -97,14 +89,30 @@ class TranzactionViewSet(viewsets.ModelViewSet):
...
+            except Exception:
+                raise Exception("404 bad ward")
```
Comment (AHTOH2001) [inline api/managment/views.py]: Replace exception with validation error

File: api/api/settings.py (hunk 1 above)
Comment (sviridchik) [reply api/api/settings.py]: noy

File: api/managment/views.py [hunk 9]
```diff
@@ -97,14 +89,30 @@ class TranzactionViewSet(viewsets.ModelViewSet):
...
+            return Doctor.objects.filter(patient__user=ward)
```
Comment (AHTOH2001) [inline api/managment/views.py]: Rewrite logic of endpoint to be able to use
`Doctor.objects.filter(patient=ward)`

File: api/managment/views.py [hunk 10]
```diff
@@ -97,14 +89,30 @@ class TranzactionViewSet(viewsets.ModelViewSet):
...
+            return Doctor.objects.filter(patient__user=self.request.user)
+
+class DoctorForWardViewSet(viewsets.ModelViewSet):
+    """ доктора подопечного (надо еще при выборе проверить что он вообще его подопечный)"""
+    serializer_class = DoctorSerializer
+    # permission_classes = [IsAuthenticated]
 
     def get_queryset(self):
-        return Doctor.objects.filter(patient__user=self.request.user)
+        return Doctor.objects.filter(patient__user=self.request.ward)
```
Comment (AHTOH2001) [inline api/managment/views.py]: Redunadant endpoint

File: api/medicine/views.py [hunk 11]
```diff
@@ -112,28 +112,48 @@ def get(self, request, pk, *args, **kwargs):
...
-    permission_classes = (IsAuthenticated,)
+    # permission_classes = (IsAuthenticated,)
 
     def get_serializer_class(self):
         if self.action in ['list', 'retrieve']:
             return ViewOnlyCureSerializer
         else:
             return MainCureSerializer
 
+    """ или лучше так ?"""
```
Comment (AHTOH2001) [inline api/medicine/views.py]: ?

**Review by AHTOH2001 — CHANGES_REQUESTED** [review]

File: api/medicine/views.py (hunk 4 above)
Comment (sviridchik) [reply api/medicine/views.py]: noy

**Comment by sviridchik** [comment]
Снимок экрана 2023-04-17 в 09 51 12

**get**
```
[
    {
        "id": 1,
        "schedule": {
            "id": 1,
            "timesheet": [
                {
                    "id": 1,
                    "time": "09:43:12"
                }
            ],
            "cycle_start": "2023-04-17",
            "cycle_end": "2023-04-18",
            "frequency": 2
        },
        "title": "qwe",
        "dose": 12.0,
        "dose_type": "ml",
        "type": "injection",
        "strict_status": false,
        "food": "While eating",
        "patient": 1
    },
    {
        "id": 2,
        "schedule": {
            "id": 1,
            "timesheet": [
                {
                    "id": 1,
                    "time": "09:43:12"
                }
            ],
            "cycle_start": "2023-04-17",
            "cycle_end": "2023-04-18",
            "frequency": 2
        },
        "title": "asd1",
        "dose": 12.0,
        "dose_type": "ml",
        "type": "injection",
        "strict_status": false,
        "food": "While eating",
        "patient": 1
    }
]
```
**{{local}}/medicine/cure/?ward=4**

**Comment by sviridchik** [comment]
Снимок экрана 2023-04-17 в 12 13 44
**POST**
{{local}}/medicine/cure/?ward=4

**Comment by sviridchik** [comment]
Снимок экрана 2023-04-17 в 12 24 38


**PATCH**
{{local}}/medicine/cure/2/?ward=4



**GET**
Снимок экрана 2023-04-17 в 12 32 18

**Comment by sviridchik** [comment]
**PUT**
{{local}}/medicine/cure/1/?ward=4
`{
    "title": "qwe",
        "dose": 123.0,
        "dose_type": "ml",
        "type": "injection",
        "food": "While eating",
        "schedule": 1
    }`

`{
    "id": 1,
    "title": "qwe",
    "dose": 123.0,
    "dose_type": "ml",
    "type": "injection",
    "schedule": 1,
    "food": "While eating",
    "strict_status": false
}`
Снимок экрана 2023-04-17 в 12 48 22


Снимок экрана 2023-04-17 в 12 52 03

**Comment by sviridchik** [comment]
**GET DOCTORS**
{{local}}/managment/doctor/?ward=4
Снимок экрана 2023-04-17 в 12 57 26

[
    {
        "id": 1,
        "first_name": "ivan",
        "last_name": "smirnov",
        "specialty": "neurologist",
        "patient": 1
    },
    {
        "id": 2,
        "first_name": "Victoria",
        "last_name": "Свиридчик",
        "specialty": "surgeon",
        "patient": 1
    },
    {
        "id": 3,
        "first_name": "test",
        "last_name": "string",
        "specialty": "therapist",
        "patient": 1
    }
]

**Comment by sviridchik** [comment]
**POST DOCTORS**
{{local}}/managment/doctor/?ward=4
Снимок экрана 2023-04-17 в 13 11 36
>
`{
    "id": 4,
    "first_name": "testPost",
    "last_name": "qwe",
    "specialty": "surgeon",
    "patient": 1
}`


`{
    "id": 5,
    "first_name": "testPost",
    "last_name": "qwe",
    "specialty": "surgeon",
    "patient": 1
}`

**Comment by sviridchik** [comment]
**PATCH**
{{local}}/managment/doctor/1/?ward=4
Снимок экрана 2023-04-17 в 14 21 05
Снимок экрана 2023-04-17 в 14 22 15
`{
    "first_name": "ivanPAtch",
    "last_name": "smirnov",
    "specialty": "neurologist",
    "patient": 1
}`



`{
    "id": 1,
    "first_name": "ivanPAtch",
    "last_name": "smirnov",
    "specialty": "neurologist",
    "patient": 1
}`

**Comment by sviridchik** [comment]
**PUT**
{{local}}/managment/doctor/1/?ward=4
Снимок экрана 2023-04-17 в 14 29 02
```
{
    "first_name": "ivanPAtch",
    "last_name": "smirnovPUT",
    "specialty": "neurologist",
    "patient": 1
}
{
    "id": 1,
    "first_name": "ivanPAtch",
    "last_name": "smirnovPUT",
    "specialty": "neurologist",
    "patient": 1
}

**Comment by sviridchik** [comment]
**GET VISIT**
{{local}}/managment/doctorvisit/?ward=4
Снимок экрана 2023-04-17 в 14 35 17
`[
    {
        "id": 1,
        "doctor": {
            "id": 1,
            "first_name": "ivanPAtch",
            "last_name": "smirnovPUT",
            "specialty": "neurologist",
            "patient": 1
        },
        "date": "2023-04-17T14:33:18+03:00",
        "patient": 1
    },
    {
        "id": 2,
        "doctor": {
            "id": 3,
            "first_name": "test",
            "last_name": "string",
            "specialty": "therapist",
            "patient": 1
        },
        "date": "2023-04-26T12:00:00+03:00",
        "patient": 1
    }
]`

**Comment by sviridchik** [comment]
**POST VISIT**
{{local}}/managment/doctorvisit/?ward=4
```
{
    "date": "2023-04-17T14:33:18+03:00",
    "doctor":  2
    
}

{
    "id": 3,
    "date": "2023-04-17T14:33:18+03:00",
    "doctor": 2,
    "patient": 1
}
```

Снимок экрана 2023-04-17 в 14 52 34
Снимок экрана 2023-04-17 в 14 54 40

**Comment by sviridchik** [comment]
**PATCH VISIT**
{{local}}/managment/doctorvisit/3/?ward=4
```
{
    "date": "2025-04-17T14:33:18+03:00",
    "doctor":  2
}
{
    "id": 3,
    "date": "2025-04-17T14:33:18+03:00",
    "doctor": 2,
    "patient": 1
}
```
Снимок экрана 2023-04-17 в 14 56 26
Снимок экрана 2023-04-17 в 14 57 39

**Comment by sviridchik** [comment]
**PUT VISIT**
{{local}}/managment/doctorvisit/3/?ward=4
```
{
    "date": "2025-10-17T14:33:18+03:00",
    "doctor":  2
}
{
    "id": 3,
    "date": "2025-10-17T14:33:18+03:00",
    "doctor": 2,
    "patient": 1
}
```

Снимок экрана 2023-04-17 в 14 59 54

Снимок экрана 2023-04-17 в 15 00 02

**Comment by sviridchik** [comment]
кто твое золотце?)

**Comment by sviridchik** [comment]
**GET ALL WARDS**
{{local}}/managment/ward/
```
[
    {
        "id": 1,
        "user": {
            "id": 4,
            "email": "patient1@gmail.com",
            "username": "patient1"
        },
        "first_name": "patient1",
        "last_name": "Sviridchik",
        "age": 41,
        "phone": 80447943822
    },
    {
        "id": 2,
        "user": {
            "id": 5,
            "email": "patient2@gmail.com",
            "username": "patient2"
        },
        "first_name": "patient2",
        "last_name": "Sviridchik",
        "age": 67,
        "phone": 80447943832
    },
    {
        "id": 3,
        "user": {
            "id": 6,
            "email": "patient3@gmail.com",
            "username": "patient3"
        },
        "first_name": "patient3",
        "last_name": "",
        "age": 98,
        "phone": 80447956822
    }
]
```
Снимок экрана 2023-04-17 в 16 01 12

**Comment by AHTOH2001** [comment]
> **PUT** {{local}}/medicine/cure/1/?ward=4 `{ "title": "qwe", "dose": 123.0, "dose_type": "ml", "type": "injection", "food": "While eating", "schedule": 1 }`
> 
> `{ "id": 1, "title": "qwe", "dose": 123.0, "dose_type": "ml", "type": "injection", "schedule": 1, "food": "While eating", "strict_status": false }` Снимок экрана 2023-04-17 в 12 48 22
> 
> Снимок экрана 2023-04-17 в 12 52 03

Allow PUT, PATCH and POST for cure modify/create schedule and timesheet

**Comment by AHTOH2001** [comment]
> **POST VISIT** {{local}}/managment/doctorvisit/?ward=4
> 
> ```
> {
>     "date": "2023-04-17T14:33:18+03:00",
>     "doctor":  2
>     
> }
> 
> {
>     "id": 3,
>     "date": "2023-04-17T14:33:18+03:00",
>     "doctor": 2,
>     "patient": 1
> }
> ```
> 
> Снимок экрана 2023-04-17 в 14 52 34
> 
> Снимок экрана 2023-04-17 в 14 54 40

Query param ward should search by patient, not by user (applies to all endpoints with ward param)

File: .idea/diplomaCurrent.iml
Comment (AHTOH2001) [inline .idea/diplomaCurrent.iml:1]: Add .idea folder to .gitignore and remove this folder from project

**Review by AHTOH2001 — CHANGES_REQUESTED** [review]

File: api/api/settings.py [hunk 12]
```diff
@@ -98,7 +98,7 @@
         'USER': 'postgres',
         'PASSWORD': 'postgres',
         'HOST': '127.0.0.1',
-        'PORT': 5432,
+        'PORT': 5433,
```
Comment (AHTOH2001) [inline api/api/settings.py]: ?

File: api/managment/utils.py
Comment (AHTOH2001) [inline api/managment/utils.py:1]: Format this file using black, plz

File: api/managment/views.py [hunk 13]
```diff
@@ -40,6 +43,21 @@ def list(self, request, *args, **kwargs):
         return Response(res, status=status.HTTP_200_OK)
 
 
+class WardViewSet(viewsets.ModelViewSet):
+    serializer_class = PatientSerializer
+    # permission_classes = [IsAuthenticated]
```
Comment (AHTOH2001) [inline api/managment/views.py]: Uncomment permission

File: api/managment/views.py [hunk 14]
```diff
@@ -40,6 +43,21 @@ def list(self, request, *args, **kwargs):
...
+
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        return queryset.get(user=self.request.user)
```
Comment (AHTOH2001) [inline api/managment/views.py:53]: Redundant code

**Review by AHTOH2001 — CHANGES_REQUESTED** [review]

**Comment by sviridchik** [comment]
**POST**

{{local}}/medicine/cure/?ward=4

Снимок экрана 2023-04-18 в 12 07 13
```

{
    "schedule": {
        "timesheet": [
            {
                "time": "19:43:12"
            },
            {
                "time": "09:43:12"
            }
            ,
            {
                "time": "09:03:12"
            }
        ],
        "cycle_start": "2023-04-17",
        "cycle_end": "2023-04-18",
        "frequency": 2
    },
    "title": "qwe",
    "dose": 123.0,
    "dose_type": "ml",
    "type": "injection",
    "strict_status": false,
    "food": "While eating",
    "patient": 1
}

{
    "id": 19,
    "title": "qwe",
    "dose": 123.0,
    "dose_type": "ml",
    "type": "injection",
    "schedule": {
        "id": 12,
        "timesheet": [
            {
                "id": 24,
                "time": "19:43:12"
            },
            {
                "id": 25,
                "time": "09:43:12"
            },
            {
                "id": 26,
                "time": "09:03:12"
            }
        ],
        "cycle_start": "2023-04-17",
        "cycle_end": "2023-04-18",
        "frequency": 2
    },
    "food": "While eating",
    "strict_status": false
}
```

**Comment by sviridchik** [comment]
**PATCH**
{{local}}/medicine/cure/19/?ward=4
Снимок экрана 2023-04-18 в 12 42 09
`	

{
    "id": 19,
    "title": "qwePATCH",
    "dose": 123.0,
    "dose_type": "ml",
    "type": "injection",
    "schedule": {
        "id": 12,
        "timesheet": [
            {
                "id": 24,
                "time": "19:43:12"
            },
            {
                "id": 25,
                "time": "09:43:12"
            },
            {
                "id": 26,
                "time": "09:03:12"
            }
        ],
        "cycle_start": "2023-04-17",
        "cycle_end": "2023-04-18",
        "frequency": 2
    },
    "food": "While eating",
    "strict_status": false
}`

`
{
    "id": 19,
    "title": "qwePATCH",
    "dose": 123.0,
    "dose_type": "ml",
    "type": "injection",
    "schedule": {
        "id": 25,
        "timesheet": [
            {
                "id": 63,
                "time": "19:43:12"
            },
            {
                "id": 64,
                "time": "09:43:12"
            },
            {
                "id": 65,
                "time": "09:03:12"
            }
        ],
        "cycle_start": "2023-04-17",
        "cycle_end": "2023-04-18",
        "frequency": 2
    },
    "food": "While eating",
    "strict_status": false
}
`
Снимок экрана 2023-04-18 в 12 43 30

**Comment by sviridchik** [comment]
**PUT**
{{local}}/medicine/cure/19/?ward=4
Снимок экрана 2023-04-18 в 12 49 43
`	

{
    "id": 19,
    "title": "qwePATCHPUTPUT",
    "dose": 123.0,
    "dose_type": "ml",
    "type": "injection",
    "schedule": {
        "id": 12,
        "timesheet": [
            {
                "id": 24,
                "time": "19:43:12"
            },
            {
                "id": 25,
                "time": "09:43:12"
            },
            {
                "id": 26,
                "time": "09:03:12"
            }
        ],
        "cycle_start": "2023-04-17",
        "cycle_end": "2023-04-18",
        "frequency": 2
    },
    "food": "While eating",
    "strict_status": false
}`

`{
    "id": 19,
    "title": "qwePATCHPUTPUT",
    "dose": 123.0,
    "dose_type": "ml",
    "type": "injection",
    "schedule": {
        "id": 26,
        "timesheet": [
            {
                "id": 66,
                "time": "19:43:12"
            },
            {
                "id": 67,
                "time": "09:43:12"
            },
            {
                "id": 68,
                "time": "09:03:12"
            }
        ],
        "cycle_start": "2023-04-17",
        "cycle_end": "2023-04-18",
        "frequency": 2
    },
    "food": "While eating",
    "strict_status": false
}`


Снимок экрана 2023-04-18 в 12 50 46

**Comment by sviridchik** [comment]
**DELETE**

Снимок экрана 2023-04-18 в 13 58 47
Снимок экрана 2023-04-18 в 13 58 57

{{local}}/medicine/cure/20/?ward=4

**Comment by sviridchik** [comment]
**DELETE**
Снимок экрана 2023-04-18 в 14 31 16
Снимок экрана 2023-04-18 в 14 31 10
{{local}}/managment/doctor/6/?ward=4

**Comment by sviridchik** [comment]
**GET VISIT WITH EXTENDED DOCTOR**

{{local}}/managment/doctorvisit/?ward=4
Снимок экрана 2023-04-18 в 14 32 09

**Comment by sviridchik** [comment]
**DELETE**
Снимок экрана 2023-04-18 в 14 37 21
Снимок экрана 2023-04-18 в 14 37 17
{{local}}/managment/doctorvisit/3/?ward=4

File: api/medicine/views.py (hunk 4 above)
Comment (sviridchik) [reply api/medicine/views.py]: noooooooooooooo, whyyyyyy

File: api/managment/views.py (hunk 8 above)
Comment (sviridchik) [reply api/managment/views.py]: There can be other exceptions

**Review by AHTOH2001 — APPROVED** [review]

**Pull request merged**